src/library/features/crates/cratehierarchy.h

```diff
@@ -0,0 +1,60 @@
+#ifndef MIXXX_CRATEHIERARCHY_H
+#define MIXXX_CRATEHIERARCHY_H
+
+#include <QObject>
+#include <QList>
+#include <QStringList>
+#include <QSet>
+
+#include "library/dao/dao.h"
+#include "library/crate/cratesummary.h"
+#include "library/features/crates/cratestoragehelpers.h"
+
+#include "util/db/fwdsqlqueryselectresult.h"
+#include "util/db/sqlsubselectmode.h"
+
+class CrateHierarchy : public virtual DAO {
+  public:
+    CrateHierarchy() {}
+    ~CrateHierarchy() override {}
+
+    void initialize(const QSqlDatabase& database) override;
+
+    ////////////////////////////////////
+    uint countCratesInClosure() const;
+
+    // empties the closure table
+    void resetClosure() const;
+    // fills the closure table with (self,self,0)
+    bool initClosure(CrateSelectResult crates) const;
```

Comment:
The word "closure" should not be in any public interface of CrateHierarchy (it's fine for private members and methods). That the hierarchy is represented in the database as a closure table is an implementation detail that only CrateHierarchy should have to know about.